### DISCUSSION TITLE
Directive Click-outside

### DISCUSSION THREAD

**Discussion by murilolivorato** [quasarframework/quasar#10995] [CLI - SSR mode]
Hello , In quasar 1 ( vue 2 ) , I used to use a directive to click-outside . but on quasar 2 , I have been trying to do it for a month , I have been no success on it . I know that on SSR you cant use el . does have any way to do it , any way thtat I can use in an application SSR as well ?

I have found this code that is a click-outside , directive in vue 3 -
https://stackoverflow.com/a/63910611/6328616

I created a boot - 
```
export default ({ app }) => {
  app.directive('click-outside', {
    beforeMount (el, binding, vnode) {
      el.clickOutsideEvent = function (event) {
        if (!(el === event.target || el.contains(event.target))) {
          binding.value(event, el)
        }
      }
      document.body.addEventListener('click', el.clickOutsideEvent)
    },
    unmounted (el) {
      document.body.removeEventListener('click', el.clickOutsideEvent)
    }
  })
}

```

I registered the boot  on conf -  
```
  boot: [
      { path: 'clickOutside', server: false }
    ],
```

and In my component 
```
<div :click-outside="outside">
methods: {
    outside () {
      console.log('clicked outside the component')
      }
}
```

**Comment by suleimanas** [comment]
with directives it should be "v-click-outside" not ":click-outside"

`<div v-click-outside="outside">abcd</div>`

**Reply by murilolivorato** [reply]
hello @suleimanas , thanks . I did the change . it is working  . But only on SPA Mode .
SSR mode is not working .  do you know what should I change to work in SSR model ? thanks

**Reply by suleimanas** [reply]
hello,
try registering the boot file without {server: false}
`boot: ["clickOutside"],`

**Reply by suleimanas** [reply]
and then you can check if ssr or not using ssrContext inside the boot export function if needed:

```
export default ({ app, ssrContext }) => {
  // ...
}
```

More:
https://quasar.dev/quasar-cli/developing-ssr/ssr-context

**Reply by murilolivorato** [reply]
@suleimanas , I dont know how to use ssrContext  , in this case .

Directives in ssr , I looks like to be different the structure -
https://quasar.dev/quasar-cli/developing-ssr/vue-ssr-directives
in the documentation is like this -
```
export default (dir) => {
  return {
    props: []
  }
}
```
It does not has any example , I didnt understand how to use it.
I must return a props , I dont know what props is this . =/ . I dont know how to access this props value from my template .

the structure that I did , is like , a spa structure  , that is the way of making directives in VUE .
https://quasar.dev/app-extensions/tips-and-tricks/provide-a-directive
```
export default ({ app }) => {
  app.directive('my-directive', MyDirective)
}
```
does some one could show me a real example of directive in SSR , some simple code that I can understand the functionality  , like this one -
```
export default (dir) => {
  return {
    props: []
  }
}
```

**Reply by rstoenescu** [reply]
Just a note that it's not our fault there's no examples of SSR directives. This is Vue docs' job. I had to implement one such directive for Quasar v2 and mainly did it by reading Vue's source code and guessing lots of stuff. At the time that I was implementing there were no examples and no mention of such a thing in Vue 3 docs website.

**Reply by murilolivorato** [reply]
quasar SSR directive , follow a different structure  -
https://quasar.dev/quasar-cli/developing-ssr/vue-ssr-directives
SOME THING LIKE THIS -
```
export default (dir) => {
  return {
    props: []
  }
}
```

**Reply by rstoenescu** [reply]
I already explained above that all directives also need to have a server-side definition as well when on SSR (otherwise Vue will fail). This definition (the exact one that you are complaining about) should be documented by Vue in its docs **because Vue is the one requiring it**! This is not something that Quasar invented (if it would have been this case then rest assured it would have been documented fully). If there is someone to complain to about the server-side directives definition, then it's the Vue team.
